File: src/policy/encoding-guard.ts
/**
 * Multi-layer encoding evasion detection — base64, hex, URL chains, and
 * mismatch between raw arguments and deobfuscated content.
 */
import type { CallContext, PolicyDecision } from './policy-types.js';
import { walkStringLeaves } from './arg-leaf-walker.js';
import { deobfuscateRecursive, stripZeroWidthCharacters } from '../utils/payload-normalizer.js';
import { detectPromptInjection } from '../scanners/prompt-injection-detector.js';
import { OVERRIDE_ATTACK_RE, SUSPICIOUS_DECODED_RE } from './encoding-guard-patterns.js';

const BASE64_BLOB_RE = /(?:^|[^A-Za-z0-9+/])([A-Za-z0-9+/]{20,}={0,2})(?:[^A-Za-z0-9+/]|$)/g;
const RAW_HEX_BLOB_RE = /\b([0-9a-fA-F]{16,})\b/g;
const PERCENT_ENCODED_RUN_RE = /(?:%[0-9a-fA-F]{2}){4,}/i;
const SHELL_FRAGMENT_RE =
  /\b(?:rm\s+-rf|union\s+select|sleep\s*\(|benchmark\s*\(|\/etc\/passwd|\/bin\/sh|select\s+\*|\/dev\/tcp|delete\s+account|drop\s+table)\b/i;

function encodingTransformApplied(original: string, decoded: string): boolean {
  if (decoded.length === 0 || decoded === original) return false;
  const norm = (s: string) => s.replace(/\r\n/g, '\n').replace(/\s+/g, ' ').trim();
  if (norm(original) === norm(decoded)) return false;
  return true;
}

/** After a real encoding transform, run keyword union + injection detector. */
function decodedContentSuspicious(original: string, decoded: string): boolean {
  if (!encodingTransformApplied(original, decoded)) return false;
  if (
    SUSPICIOUS_DECODED_RE.test(decoded)
    || OVERRIDE_ATTACK_RE.test(decoded)
    || SHELL_FRAGMENT_RE.test(decoded)
  ) {
    return true;
  }
  const findings = detectPromptInjection('encoding-guard', decoded);
  return findings.some((f) => f.severity === 'critical' || f.severity === 'high');
}

export function isEncodingGuardEnabled(): boolean {
  return process.env['MASTYFF_AI_ENCODING_GUARD'] !== 'false';
}

function tryDecodeBase64(b64: string): string | null {
  if (b64.length < 12 || b64.length % 4 === 1) return null;
  try {
    const decoded = Buffer.from(b64, 'base64').toString('utf8');
    if (decoded.length < 4 || !/^[\x20-\x7E\u00A0-\uFFFF\s]+$/.test(decoded)) return null;
    return decoded;
  } catch {
    return null;
  }
}

function tryDecodeRawHex(hex: string): string | null {
  if (hex.length < 16 || hex.length % 2 !== 0) return null;
  try {
    const decoded = Buffer.from(hex, 'hex').toString('utf8');
    if (decoded.length < 4 || !/^[\x20-\x7E]+$/.test(decoded)) return null;
    return decoded;
  } catch {
    return null;
  }
}

export function scanEncodingEvasion(blob: string): { matched: boolean; reason: string } {
  if (!blob.trim()) return { matched: false, reason: '' };

  const deobfuscated = deobfuscateRecursive(blob);
  const strippedInvisible = stripZeroWidthCharacters(blob);
  if (decodedContentSuspicious(blob, deobfuscated)) {
    return { matched: true, reason: 'multi-layer encoding reveals blocked content after decode' };
  }
  if (strippedInvisible !== blob && decodedContentSuspicious(blob, strippedInvisible)) {
    return { matched: true, reason: 'zero-width stripped content is suspicious after decode' };
  }

  if (PERCENT_ENCODED_RUN_RE.test(blob) && decodedContentSuspicious(blob, deobfuscated)) {
    return { matched: true, reason: 'percent-encoded payload decodes to suspicious content' };
  }

  for (const match of blob.matchAll(BASE64_BLOB_RE)) {
    const b64 = match[1];
    const decoded = tryDecodeBase64(b64);
    if (decoded && decodedContentSuspicious(b64, decoded)) {
      return { matched: true, reason: 'base64 blob decodes to suspicious instruction text' };
    }
  }

  for (const match of blob.matchAll(RAW_HEX_BLOB_RE)) {
    const hex = match[1];
    const decoded = tryDecodeRawHex(hex);
    if (decoded && decodedContentSuspicious(hex, decoded)) {
      return { matched: true, reason: 'raw hex blob decodes to suspicious instruction text' };
    }
  }

  const trimmed = blob.trim();
  if (/^[A-Za-z0-9+/]+={0,2}$/.test(trimmed)) {
    const whole = tryDecodeBase64(trimmed);
    if (whole && decodedContentSuspicious(trimmed, whole)) {
      return { matched: true, reason: 'whole-string base64 decodes to suspicious content' };
    }
  }

  return { matched: false, reason: '' };
}

export function evaluateEncodingGuard(ctx: CallContext): PolicyDecision | null {
  if (!isEncodingGuardEnabled()) return null;

  const blob = walkStringLeaves(ctx.arguments ?? {})
    .map((l) => l.value)
    .join('\n');
  if (!blob.trim()) return null;

  const scan = scanEncodingEvasion(blob);
  if (!scan.matched) return null;

  return {
    action: 'block',
    rule: 'encoding-evasion-guard',
    reason: scan.reason,
  };
}

File: adversarial-harness/python/policy_engine/encoding_guard.py
"""Encoding evasion detection — mirrors encoding-guard.ts."""

from __future__ import annotations

import base64
import os
import re
from typing import Optional

from .arg_walker import walk_string_leaves
from .normalizer import deobfuscate_recursive, ZERO_WIDTH_RE
from .types import CallContext, PolicyDecision

BASE64_BLOB_RE = re.compile(r"(?:^|[^A-Za-z0-9+/])([A-Za-z0-9+/]{20,}={0,2})(?:[^A-Za-z0-9+/]|$)")
RAW_HEX_BLOB_RE = re.compile(r"\b([0-9a-fA-F]{16,})\b")
PERCENT_ENCODED_RUN_RE = re.compile(r"(?:%[0-9a-fA-F]{2}){4,}", re.I)
SUSPICIOUS_DECODED_RE = re.compile(
    r"\b(?:ignore|disregard|bypass|jailbreak|delete|drop|exec|eval|curl|wget|"
    r"rm\s+-rf|union\s+select|sleep\s*\(|benchmark\s*\(|/etc/passwd|bash|/bin/sh|"
    r"select\s+\*|\bselect\b|/dev/tcp)\b",
    re.I,
)
OVERRIDE_ATTACK_RE = re.compile(
    r"\boverride\b.{0,80}\b(?:all|previous|prior|safety|instruction|rules|system|filter|restriction|guidance)\b",
    re.I,
)


def _decoded_suspicious(text: str) -> bool:
    return bool(SUSPICIOUS_DECODED_RE.search(text) or OVERRIDE_ATTACK_RE.search(text))


def is_encoding_guard_enabled() -> bool:
    return os.environ.get("MASTYFF_AI_ENCODING_GUARD") != "false"


def _try_decode_base64(b64: str) -> Optional[str]:
    if len(b64) < 12 or len(b64) % 4 == 1:
        return None
    try:
        decoded = base64.b64decode(b64 + "==="[: (4 - len(b64) % 4) % 4], validate=False).decode("utf-8")
        if len(decoded) < 4:
            return None
        return decoded
    except Exception:
        return None


def _try_decode_raw_hex(hex_str: str) -> Optional[str]:
    if len(hex_str) < 16 or len(hex_str) % 2 != 0:
        return None
    try:
        decoded = bytes.fromhex(hex_str).decode("utf-8")
        if len(decoded) < 4 or not all(0x20 <= ord(c) <= 0x7E for c in decoded):
            return None
        return decoded
    except Exception:
        return None


def scan_encoding_evasion(blob: str) -> tuple[bool, str]:
    if not blob.strip():
        return False, ""
    deobfuscated = deobfuscate_recursive(blob)
    stripped_invisible = ZERO_WIDTH_RE.sub("", blob)
    if deobfuscated and _decoded_suspicious(deobfuscated):
        if deobfuscated != blob or (
            stripped_invisible != blob and _decoded_suspicious(stripped_invisible)
        ):
            return True, "multi-layer encoding reveals blocked content after decode"
    if PERCENT_ENCODED_RUN_RE.search(blob) and _decoded_suspicious(deobfuscated):
        return True, "percent-encoded payload decodes to suspicious content"
    for match in BASE64_BLOB_RE.finditer(blob):
        decoded = _try_decode_base64(match.group(1))
        if decoded and _decoded_suspicious(decoded):
            return True, "base64 blob decodes to suspicious instruction text"
    for match in RAW_HEX_BLOB_RE.finditer(blob):
        decoded = _try_decode_raw_hex(match.group(1))
        if decoded and _decoded_suspicious(decoded):
            return True, "raw hex blob decodes to suspicious instruction text"
    trimmed = blob.strip()
    if re.fullmatch(r"[A-Za-z0-9+/]+={0,2}", trimmed):
        whole = _try_decode_base64(trimmed)
        if whole and _decoded_suspicious(whole):
            return True, "whole-string base64 decodes to suspicious content"
    return False, ""


def evaluate_encoding_guard(ctx: CallContext) -> Optional[PolicyDecision]:
    if not is_encoding_guard_enabled():
        return None
    blob = "\n".join(leaf.value for leaf in walk_string_leaves(ctx.arguments or {}))
    if not blob.strip():
        return None
    matched, reason = scan_encoding_evasion(blob)
    if not matched:
        return None
    return PolicyDecision("block", "encoding-evasion-guard", reason)
